x/fundraising/keeper/match.go

```diff
@@ -46,140 +47,56 @@ func (k Keeper) CalculateFixedPriceAllocation(ctx sdk.Context, auction types.Auc
-					var reserveAmt sdk.Int
-					if matchingAmt.LT(bidAmt) {
-						reserveAmt = matchingAmt.ToDec().Mul(matchingPrice).Ceil().TruncateInt()
-					} else {
-						reserveAmt = b.Coin.Amount
-					}
-					reservedMatchedMap[b.Bidder] = reservedMatchedAmt.Add(reserveAmt)
-				}
-
-				totalMatchedAmt = totalMatchedAmt.Add(matchingAmt)
-			} else {
-				bidAmt := b.Coin.Amount
-
-				// MinInt(BidAmt, MaxBidAmount-AccumulatedBidAmount)
-				matchingAmt := sdk.MinInt(bidAmt, maxBidAmt.Sub(allocateAmt))
-
-				// Accumulate matching amount since a bidder can have multiple bids
-				if alloc, ok := allocationMap[b.Bidder]; ok {
-					allocationMap[b.Bidder] = alloc.Add(matchingAmt)
-				}
-
-				// Accumulate how much reserved paying coin amount is matched
-				if reservedMatchedAmt, ok := reservedMatchedMap[b.Bidder]; ok {
-					reserveAmt := matchingAmt.ToDec().Mul(matchingPrice).Ceil().TruncateInt()
-					reservedMatchedMap[b.Bidder] = reservedMatchedAmt.Add(reserveAmt)
-				}
-
-				totalMatchedAmt = totalMatchedAmt.Add(matchingAmt)
-			}
+	var matchRes *types.MatchResult
+	for i, price := range prices {
+		res, found := types.Match(auction, price, prices, bidsByPrice)
+		if found || (matchRes == nil && i == len(prices)-1) {
+			matchRes = res
 		}
-
-		// Exit the iteration when the total matched amount is greater than the total selling coin amount
-		if totalMatchedAmt.GT(auction.GetSellingCoin().Amount) {
+		if !found {
 			break
 		}
+	}
 
-		mInfo.MatchedLen = mInfo.MatchedLen + 1
-		mInfo.MatchedPrice = matchingPrice
-		mInfo.TotalMatchedAmount = totalMatchedAmt
+	matchingInfo.MatchedLen = int64(len(matchRes.MatchedBids))
+	matchingInfo.MatchedPrice = matchRes.MatchPrice
+	matchingInfo.TotalMatchedAmount = matchRes.MatchedAmount
 
-		for _, ab := range auction.GetAllowedBidders() {
-			mInfo.AllocationMap[ab.Bidder] = allocationMap[ab.Bidder]
-			mInfo.ReservedMatchedMap[ab.Bidder] = reservedMatchedMap[ab.Bidder]
+	reservedAmtByBidder := map[string]sdk.Int{}
+	for _, bid := range bids {
+		bidderReservedAmt, ok := reservedAmtByBidder[bid.Bidder]
+		if !ok {
+			bidderReservedAmt = sdk.ZeroInt()
 		}
-
-		bid.SetMatched(true)
-		k.SetBid(ctx, bid)
+		reservedAmtByBidder[bid.Bidder] = bidderReservedAmt.Add(bid.ConvertToPayingAmount(auction.GetPayingCoinDenom()))
 	}
 
-	// Iterate all bids to get refund amount for each bidder
-	// Calculate the refund amount by substracting allocate amount from
-	// how much a bidder reserved to place a bid for the auction
-	for _, b := range bids {
-		if b.Type == types.BidTypeBatchWorth {
-			reservedMap[b.Bidder] = reservedMap[b.Bidder].Add(b.Coin.Amount)
-		} else {
-			bidAmt := b.Coin.Amount.ToDec().Mul(b.Price).Ceil().TruncateInt()
-			reservedMap[b.Bidder] = reservedMap[b.Bidder].Add(bidAmt)
-		}
+	for bidder, reservedAmt := range reservedAmtByBidder {
+		matchingInfo.AllocationMap[bidder] = sdk.ZeroInt()
+		matchingInfo.ReservedMatchedMap[bidder] = sdk.ZeroInt()
```

Comment:
Do we initialize these here for the next for loop of `matchRes.MatchResultByBidder`?

Reply:
Bidders who have reserved their paying coin could be not-matched, so we initialize `AllocationMap` and `ReservedMatchedMap` to zero for all bidders.